cmd/goquery_completion/conditional.go

```diff
@@ -169,3 +285,127 @@ func conditional(args []string) []string {
+		dirKeywordCount > 0 {
+		// filter out invalid suggestions
+		if !checkDirKeywordConstraints(sugg, tokens, prev, openParens, dirKeywordCount) {
+			return false
+		}
+	}
+	return true
+}
+
+// checkDirKeywordConstraints checks whether sugg is a valid suggestion with respect to the
+// constraints on the structure of the condition string introduced by the dir keyword.
+// Returns true if sugg is a valid suggestion and false otherwise.
+func checkDirKeywordConstraints(sugg suggestion, tokens []string, prev string, openParens, dirKeywordCount int) bool {
+	// determine whether prev is a top-level conjunction
+	firstUnnestedBinaryOpPos, firstUnnestedBinaryOp := firstUnnestedBinaryOp(tokens)
+	nTokens := len(tokens)
+	topLevelAnd := firstUnnestedBinaryOp == "&" && firstUnnestedBinaryOpPos == nTokens-2
+	// determine whether a top-level conjunction has previously occurred
+	topLevelAndOccurred := firstUnnestedBinaryOp == "&" && firstUnnestedBinaryOpPos <= nTokens-2
+
+	// the dir keyword must only be suggested at the start of the condition (condition string currently empty)
+	// or directly after a top-level conjunction
+	if !strings.Contains(sugg.token, "directional") &&
+		(strings.Contains(sugg.token, types.FilterKeywordDirection) || strings.Contains(sugg.token, types.FilterKeywordDirectionSugared)) {
+		if !(prev == "" || (topLevelAnd && dirKeywordCount == 0)) {
+			return false
+		}
+	}
+
+	// an 'unnested' disjunction is disallowed if the dir keyword is already present
+	if sugg.token == "|" && dirKeywordCount > 0 && openParens == 0 {
+		return false
+	}
+
+	// if dir keyword has already occurred and there is already a top-level conjunction,
+	// no other 'unnested' conjunction or disjunction is allowed
+	//(otherwise dir keyword would not be part of the top-level conjunct anymore)
+	if sugg.token == "&" && dirKeywordCount > 0 && topLevelAndOccurred && openParens == 0 {
+		return false
+	}
+
+	return true
+}
+
+// conditionMustTerminate checks whether the condition string must end now in order to be a valid condition.
+// Returns true if condition string must be terminated and false otherwise.
+func conditionMustTerminate(suggs []suggestion, prevprev, prev, last string, openParens, dirKeywordCount, nTokens int) bool {
+	// If there are no suggestions, and there is already a previous condition,
+	// and the current condition is a direction filter, the condition string
+	// must end (no other conditions are allowed afterwards).
+	//
+	// Note: this case handles a dir keyword occurring
+	// on the right of a top-level conjunction (potentially
+	// with trailing whitespaces)
+	// (e.g. terminates "sip = 127.0.0.1 & dir = in ",
+	// "sip = 127.0.0.1 & dir = inbound ")
+	if len(suggs) == 0 && nTokens > 3 {
+		for _, direction := range types.DirectionFilters {
+			// if dir is specified on the right side of a top-level conjunction,
+			// do not allow any trailing tokens
+			if prev == string(direction) && last == "" || last == string(direction) {
```

Comment:
Just an idea: You could change the definition of the `FilterTypeDirection` in `types.go` to:
```
// FilterTypeDirection denotes filters wrt. directionality of traffic
type FilterTypeDirection = string
```

That way, you could skip the casts for comparisons everywhere. On the other hand you'd weaken / loosen the typing, but having to cast for string comparisons is often an indicator that you actually want the type equality... But as I said, just an idea...

Reply:
I think for now your suggestion is the cleanest solution and I adjusted the code accordingly. 
In case other types of filters are needed in the future, it might be nice to separate the different kinds of filters via the type system.